File: src/algorithms/arrays/two-pointer/merge-sorted-arrays/sources/merge-sorted-arrays_test.go
package mergesortedarrays

import (
	"reflect"
	"testing"
)

func TestBasicMerge(t *testing.T) {
	result := mergeSortedArrays([]int{1, 3, 5}, []int{2, 4, 6})
	expected := []int{1, 2, 3, 4, 5, 6}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestEmptyFirstArray(t *testing.T) {
	result := mergeSortedArrays([]int{}, []int{1, 2, 3})
	expected := []int{1, 2, 3}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestEmptySecondArray(t *testing.T) {
	result := mergeSortedArrays([]int{1, 2, 3}, []int{})
	expected := []int{1, 2, 3}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestBothEmpty(t *testing.T) {
	result := mergeSortedArrays([]int{}, []int{})
	if len(result) != 0 {
		t.Errorf("Expected empty, got %v", result)
	}
}

func TestOverlappingValues(t *testing.T) {
	result := mergeSortedArrays([]int{1, 2, 4}, []int{2, 3, 5})
	expected := []int{1, 2, 2, 3, 4, 5}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestSingleElementArrays(t *testing.T) {
	result := mergeSortedArrays([]int{5}, []int{3})
	expected := []int{3, 5}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestDefaultInput(t *testing.T) {
	result := mergeSortedArrays([]int{1, 3, 5, 7, 9}, []int{2, 4, 6, 8, 10})
	expected := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

File: src/algorithms/arrays/two-pointer/merge-sorted-arrays/__tests__/merge-sorted-arrays_test.rs
include!("../sources/merge-sorted-arrays.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_merge() {
        let result = merge_sorted_arrays(&[1, 3, 5], &[2, 4, 6]);
        assert_eq!(result, vec![1, 2, 3, 4, 5, 6]);
    }

    #[test]
    fn test_empty_first_array() {
        let result = merge_sorted_arrays(&[], &[1, 2, 3]);
        assert_eq!(result, vec![1, 2, 3]);
    }

    #[test]
    fn test_empty_second_array() {
        let result = merge_sorted_arrays(&[1, 2, 3], &[]);
        assert_eq!(result, vec![1, 2, 3]);
    }

    #[test]
    fn test_both_empty() {
        let result = merge_sorted_arrays(&[], &[]);
        assert_eq!(result, vec![]);
    }

    #[test]
    fn test_overlapping_values() {
        let result = merge_sorted_arrays(&[1, 2, 4], &[2, 3, 5]);
        assert_eq!(result, vec![1, 2, 2, 3, 4, 5]);
    }

    #[test]
    fn test_single_element_arrays() {
        let result = merge_sorted_arrays(&[5], &[3]);
        assert_eq!(result, vec![3, 5]);
    }

    #[test]
    fn test_different_lengths() {
        let result = merge_sorted_arrays(&[1, 10], &[2, 3, 4, 5, 6]);
        assert_eq!(result, vec![1, 2, 3, 4, 5, 6, 10]);
    }

    #[test]
    fn test_default_input() {
        let result = merge_sorted_arrays(&[1, 3, 5, 7, 9], &[2, 4, 6, 8, 10]);
        assert_eq!(result, vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    }
}
